benchmark-framework/src/main/java/io/openmessaging/benchmark/WorkloadGenerator.java

```diff
@@ -265,26 +300,77 @@ public void close() throws Exception {
-            RateLimiter rateLimiter) {
-        Recorder recorder = new Recorder(TimeUnit.SECONDS.toMicros(30), 5);
-        Recorder cumulativeRecorder = new Recorder(TimeUnit.SECONDS.toMicros(30), 5);
+    // Results for consumer are recorded via the callback
+    private void generateConsumerLoad(List<BenchmarkConsumer> consumers, RateLimiter rateLimiter) {
+        executor.submit(() -> {
+            try {
+                // Consume messages on all topics/producers
+                while (!testCompleted) {
+                    consumers.forEach(consumer -> {
+                        rateLimiter.acquire();
+                        consumer.receiveAsync(this).exceptionally(ex -> {
+                            log.warn("Write error on message", ex);
+                            System.exit(-1);
+                            return null;
+                        });
+                    });
+                }
+            } catch (Throwable t) {
+                log.error("Got error", t);
+            }
+        });
+    }
 
-        long startTime = System.nanoTime();
-        this.testCompleted = false;
+    /**
+     * Generate a string that is sufficently large so that we can build messages on uniform text for validation purposes.
+     *
+     * @return
+     */
+    private byte[] getComparisonStr(int sizeBytes) {
+        StringBuilder stringBuilder = new StringBuilder();
+        char[] alphabet = "abcdefghijklmnopqrstuvwxyz".toCharArray();
+        // Message number (long) + producer hashcode (int) + createTimestamp (long) + trim last char (char)
+        int overhead = 8 + 4 + 8;
+        int sizeMinusOverhead = sizeBytes - overhead;
+        for (int i = 0; i < sizeMinusOverhead; i++) {
+            stringBuilder.append(alphabet[i%(alphabet.length-1)]);
+        }
+        return stringBuilder.toString().getBytes();
+    }
+
+    private void generateProducerLoad(List<BenchmarkProducer> producers, RateLimiter rateLimiter,
+                                            Recorder recorder, Recorder cumulativeRecorder) {
+        if (producers.size() == 0) {
+            log.info("Skipping producer load generation because no producers were requested");
+            return;
+        }
+        byte[] comparisonStr = getComparisonStr(workload.messageSize);
 
         executor.submit(() -> {
             try {
-                byte[] payloadData = new byte[workload.messageSize];
-
                 // Send messages on all topics/producers
+                AtomicLong messageNumber = new AtomicLong(totalMessagesSent.longValue() + 1);
                 while (!testCompleted) {
-                    for (int i = 0; i < producers.size(); i++) {
-                        BenchmarkProducer producer = producers.get(i);
+                    for (BenchmarkProducer producer : producers) {
                         rateLimiter.acquire();
 
+                        final byte[] payloadData = new byte[workload.messageSize];
```

Comment:
All these `byte[]` are going to generate a lot of mem allocations in the load generator

Reply:
Good point, I should pull this out-side the while-loop along with population of `comparisonStr`; that way, each message only needs to fill in the first 20 bytes.